mathematica/ieee754_floating_point_evaluation.wl

```diff
@@ -141,27 +160,137 @@ Block[
 evae[a_*b_+c_]:=If[
 usefma,
-applyOp[#1 #2+#3&,evae[a],evae[b],evae[c]],
-applyOp[Plus,applyOp[Times,evae[a],evae[b]],evae[c]]];
-evae[a_+b_]:=applyOp[Plus,evae[a],evae[b]];
+applyOpWithAbsoluteError[#1 #2+#3&,evae[a],evae[b],evae[c]],
+applyOpWithAbsoluteError[Plus,applyOpWithAbsoluteError[Times,evae[a],evae[b]],evae[c]]];
+evae[a_+b_]:=applyOpWithAbsoluteError[Plus,evae[a],evae[b]];
 evae[a_+b__]:=(Message[IEEEEvaluateWithAbsoluteError::badass]; $Failed);
-evae[a_*b_]:=applyOp[Times,evae[a],evae[b]];
+evae[a_*b_]:=applyOpWithAbsoluteError[Times,evae[a],evae[b]];
 evae[a_*b__]:=(Message[IEEEEvaluateWithAbsoluteError::badass]; $Failed);
-evae[a_/b_]:=applyOp[Divide,evae[a],evae[b]];
+evae[a_/b_]:=applyOpWithAbsoluteError[Divide,evae[a],evae[b]];
 (* Negation is exact. *)
 evae[-a_]:=-evae[a];
 (* Squaring an interval is not the same as multiplying two identical intervals. *) 
-evae[a_^2]:=applyOp[#^2&,evae[a]];
-evae[a_^3]:=applyOp[cube,evae[a]];
-evae[a_^4]:=applyOp[#^2&,applyOp[#^2&,evae[a]]];
+evae[a_^2]:=applyOpWithAbsoluteError[#^2&,evae[a]];
+evae[a_^3]:=applyOpWithAbsoluteError[cube,evae[a]];
+evae[a_^4]:=applyOpWithAbsoluteError[#^2&,applyOpWithAbsoluteError[#^2&,evae[a]]];
 evae[a_?NumberQ]:=Block[{cra=CorrectlyRound[a]},evae[Interval[{cra,cra}]]];
 evae[{v_Interval,\[Delta]_Interval}]:={v,\[Delta]};
 evae[a_Interval]:={a,Interval[{0,0}]};
 evae[a_?ValueQ]:=evae[Evaluate[a]];
 evae[a_]:=a;
 evae[x]];
 IEEEEvaluateWithAbsoluteError[_, args__]:=
-(Message[IEEEEvaluate::argnum, Length[{args}] + 1]; $Failed);
+(Message[IEEEEvaluateWithAbsoluteError::argnum, Length[{args}] + 1]; $Failed);
+
+
+(* ::Text:: *)
+(*The relative error bound on an IEEE computation.*)
+
+
+relativeErrorBound := If[
+	RoundingMode[]==NearestTiesToEven,
+	FromRepresentation[Representation[1]+1/2]-1,
+	FromRepresentation[Representation[1]+1]-1];
+
+
+(* ::Text:: *)
+(*Returns the relative error bound for a computation.  The relative error is "small", i.e., close to 2^-53, not to (1+2^-53), and is an interval.  The arguments are pairs of {value interval, relative error interval}.*)
+
+
+applyOpWithRelativeError[square,{va_,\[Delta]a_}]:=Block[
+	{h,r,\[Delta]r},
+	r=va^2;
+	r=Interval[{CorrectlyRound[Min[r]],CorrectlyRound[Max[r]]}];
+	h=relativeErrorBound;
+	\[Delta]r=(1+\[Delta]a)^2 Interval[{1-h,1+h}]-1;
+	{r,\[Delta]r}];
+applyOpWithRelativeError[cube,{va_,\[Delta]a_}]:=Block[
+	{h,r,\[Delta]r},
+	r=va^3;
+	h=relativeErrorBound;
+	\[Delta]r=(1+\[Delta]a)^3 Interval[{1-h,1+h}] Interval[{1-h,1+h}]-1;
+	{r,\[Delta]r}];
+applyOpWithRelativeError[fourth,{va_,\[Delta]a_}]:=Block[
+	{h,r,\[Delta]r},
+	r=va^4;
+	h=relativeErrorBound;
+	\[Delta]r=(1+\[Delta]a)^4 Interval[{1-h,1+h}]Interval[{1-h,1+h}]-1;
+	{r,\[Delta]r}];
+applyOpWithRelativeError[Plus,{va_,\[Delta]a_},{vb_,\[Delta]b_}]:=Block[
+	{corners,err,h,\[Theta]2,\[Theta]\[Prime]2,r,\[Delta]r},
+	r=va+vb;
+	r=Interval[{CorrectlyRound[Min[r]],CorrectlyRound[Max[r]]}];
+	If[
+		IntervalMemberQ[r,0],
```

Comment:
```suggestion
	{corners,err,h,\[Theta]2,\[Theta]\[Prime]2,r,\[Delta]r,nonZero,rNonZero},
	r=va+vb;
	r=Interval[{CorrectlyRound[Min[r]],CorrectlyRound[Max[r]]}];
	nonZero=Interval[{-\[Infinity],-FromRepresentation[1]},{FromRepresentation[1],\[Infinity]}];
	rNonZero=IntervalIntersection[nonzero,va]+IntervalIntersection[nonzero,vb];
	rNonZero=Interval[{CorrectlyRound[Min[rNonZero]],CorrectlyRound[Max[rNonZero]]}];
	If[
		IntervalMemberQ[rNonZero,0],
```
And add a test about `IEEEEvaluateWithRelativeError[Interval[{0,1}]+Interval[{0,1}]]`.